lib/StackAnalysis/SegregateDirectStackAccesses.cpp

```diff
@@ -0,0 +1,152 @@
+
+  for (BasicBlock &BB : F) {
+    for (Instruction &I : BB) {
+      Value *Ptr = nullptr;
+
+      // Differentiate accesses and add them onto their respective bucket.
+      // Everything that is not a direct access on the stack is put onto the
+      // bucket `NotDirectStackAccesses`. Load/store that access the CSVs will
+      // have their alias info added later as well.
+      if (auto *LI = dyn_cast<LoadInst>(&I))
+        Ptr = skipCasts(LI->getPointerOperand());
+      else if (auto *SI = dyn_cast<StoreInst>(&I))
+        Ptr = skipCasts(SI->getPointerOperand());
+
+      if (Ptr) {
+        Value *LoadPtr, *LHS, *BitCast;
+        ConstantInt *Offset;
+        Builder.SetInsertPoint(&I);
+
+        // Do we have a inttoptr to a load i64, i64* LoadPtr as pointer operand
+        // of the current instruction, where LoadPtr is SP? Change it with the
+        // newly-created bitcasted load in order to prevent from using inttoptr.
+        if (match(Ptr, m_Load(m_Value(LoadPtr)))
+            && GCBI->isSPReg(skipCasts(LoadPtr))) {
+          BitCast = Builder.CreateBitCast(SPI8Ptr,
+                                          isa<LoadInst>(&I) ?
+                                            I.getOperand(0)->getType() :
+                                            I.getOperand(0)
+                                              ->getType()
+                                              ->getPointerTo());
+          I.setOperand(isa<LoadInst>(&I) ? 0 : 1, BitCast);
+
+          DirectStackAccesses.emplace_back(&I);
+
+          // Do we have a inttoptr to a add i64 LHS, -8 as pointer operand of
+          // the current instruction, where LHS is SP? Canonicalize the i2p +
+          // add into a gep whose value is bitcasted to the original type of SP.
+        } else if (match(Ptr, m_c_Add(m_Value(LHS), m_ConstantInt(Offset)))
+                   && LHS == LoadSP) {
+          auto *GEP = Builder.CreateGEP(Builder.getInt8Ty(), SPI8Ptr, Offset);
+          BitCast = Builder.CreateBitCast(GEP,
+                                          isa<LoadInst>(&I) ?
+                                            I.getOperand(0)->getType() :
+                                            I.getOperand(0)
+                                              ->getType()
+                                              ->getPointerTo());
+          I.setOperand(isa<LoadInst>(&I) ? 0 : 1, BitCast);
+
+          DirectStackAccesses.emplace_back(&I);
+        } else {
+          NotDirectStackAccesses.emplace_back(&I);
+        }
+      }
+    }
+  }
+}
+
+void SDSAP::decorateStackAccesses() {
+  MDBuilder MDB(*Context);
+
+  MDNode *AliasDomain = MDB.createAliasScopeDomain("CSVAliasDomain");
+
+  auto *DSAScope = MDB.createAliasScope("DirectStackAccessScope", AliasDomain);
+  auto *NDSAScope = MDB.createAliasScope("Not(DirectStackAccessScope)",
+                                         AliasDomain);
```

Comment:
Maybe add a comment on how `Not(DirectStackAccessScope)` is different from `NotDirectStackAccessScope`.